### ISSUE TITLE
Provide a public interface to reset a Starlette application

### ISSUE THREAD

**Issue by Lawouach** [Kludex/starlette#926]
Hi,

I had been struggling with a very odd issue for a long time. I have a fairly long test suite which I have paramaterized so its iterations grows linearly with the number of parameters.

For some time, when I extend that list from two elements to three (or more) some of my tests would systematically fail with a "maximum recursion depth" error. Always the same tests. But they would never fail when executed on their own.

Initially I was so puzzled because they failed in odd places until I debugged a bit and found the sesison middleware was in an infinite loop between its send/send_wrapper.

But there was nothing that could justify this in the code.

After hours of investigating, I finally found the issue.

My routing is made of of various levels of routing endpoint instances to a root level starlette app. However, two legacy apps are also mounted to that root app. Both have their own list of middlewares.

Those two apps are two regular Starlette app instanciated in their own Python module. The typical:

```
api = Starlette()

@api.route("/")
async def homepage(request):
    return JSONResponse({'hello': 'world'})
```

What I had failed to appreciate is the singleton nature of these apps. Meaning their definition was shared all across my tests.

In effect, that led to their middlewares to be never cleared out but kept throughout.

As my number of tests grew, I was getting closer to the recursion limit...

It took me a long time down the rabbit hole to understand what to do: clear the list of user middlware before starting a new test session so previous middlewares would not be kept.

Here is an example:

app.py
```python
from starlette.applications import Starlette
from starlette.middleware.gzip import GZipMiddleware
from starlette.responses import JSONResponse
from starlette.routing import Mount

api = Starlette()


@api.route("/")
async def homepage(request):
    return JSONResponse({'hello': 'world'})


def create_app() -> Starlette:
    api.add_middleware(GZipMiddleware)
    return Starlette(debug=True, routes=[
        Mount('/', app=api)
    ])
```

conftest.py
```python
from .app import create_app

import pytest
from starlette.applications import Starlette


@pytest.fixture(params=["english", "french"])
def app() -> Starlette:
    return create_app()
```

test.py
```python
from starlette.applications import Starlette

from .app import api


def test_duplicate_middlewares(app: Starlette):
    print(app.user_middleware)
    print(api.user_middleware)
```

```console
$ pytest.py -s test.py 
Test session starts (platform: linux, Python 3.7.5, pytest 5.3.5, pytest-sugar 0.9.2)
rootdir: .
plugins: asyncio-0.10.0, sugar-0.9.2, requests-mock-1.7.0, cov-2.8.1, pylama-7.7.1
collecting ... []
[Middleware(GZipMiddleware)]

 test.py ✓                                                                                                                                                                                             50% █████     []
[Middleware(GZipMiddleware), Middleware(GZipMiddleware)]
 test.py ✓✓                                                                                                                                                                                           100% ██████████

Results (0.04s):
       2 passed
```

Two tests, one for each value of the parameter list.

To fix this:

```python
from .app import api, create_app

import pytest
from starlette.applications import Starlette


@pytest.fixture(params=["english", "french"])
def app() -> Starlette:
    api.user_middleware = []
    api.middleware_stack = api.build_middleware_stack()
    return create_app()
```

```console
Test session starts (platform: linux, Python 3.7.5, pytest 5.3.5, pytest-sugar 0.9.2)
rootdir: .
plugins: asyncio-0.10.0, sugar-0.9.2, requests-mock-1.7.0, cov-2.8.1, pylama-7.7.1
collecting ... []
[Middleware(GZipMiddleware)]

 test.py ✓                                                                                                                                                                                             50% █████     []
[Middleware(GZipMiddleware)]
 test.py ✓✓                                                                                                                                                                                           100% ██████████

Results (0.04s):
       2 passed
```

Strictly speaking the `api.middleware_stack = api.build_middleware_stack()` is not useful in this dummy example but was necessary in more complex setup.

So, while I understand the implication of a singleton and the fact, when running, the app will not see this duplication, I would like to suggest an official public `reset app` method on the Starlette interface for testing purpose. Or at least, understand if resetting middlewares as I do is enough.

Thanks,
Sylvain

**Comment by Lawouach** [comment]
An explanation would have been civil and appreciated.

**Comment by lovelydinosaur** [comment]
Up for chatting this over on the discussion group or similar, but we’re not doing anything differently here to any other ASGI or WSGI app in the Python ecosystem.

We don’t document the decorator syntax for middleware anymore, in favour of the pass-it-to-the-app-instance style, which I’d guess makes things clearer here anyways?

**Comment by Lawouach** [comment]
The passing of middlewares as part of the `api = Starlette()` call would be complicated in large applications. The api instance is created at module level, I dont always know all my list of middlewares then. I need to postpone and use `add_middleware`.

**Comment by lovelydinosaur** [comment]
> explanation would have been

Heya. It’s all good, I was just in the process of typing out a response.

I’m completely swamped by the volume of issues that come in across all the encode projects, so have been trying to encourage folks to start conversations in the discussion group first as the “support ticket” nature of GitHubs issues interface is really draining.

If some stuff gets closed quickly, it’s not a personal comment. It’s just trying to manage things succinctly, and not getting overwhelmed. 💚


https://github.com/encode/.github/blob/master/CONTRIBUTING.md

**Comment by Lawouach** [comment]
Ah I see. Thanks for clarifying. GH should have a "close but with love" button.

I will try better next time.
